ICE: field should be captured by immutable ref if we have an `Fn` instance

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![feature(async_closure)]

fn main() {}

fn needs_fn_mut<T>(mut x: impl FnMut() -> T) {
    x();
}

fn hello(x: Ty) {
    needs_fn_mut(async || {
        x.hello();
    });
}

struct Ty;
impl Ty {
    fn hello(self) {}
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.82.0-nightly (804be74e3 2024-08-19)
binary: rustc
commit-hash: 804be74e3c4ec3646b448dfdb7a0fc31237a5818
commit-date: 2024-08-19
host: x86_64-unknown-linux-gnu
release: 1.82.0-nightly
LLVM version: 19.1.0
```
`rustc -Zvalidate-mir  --edition=2018 --crate-type=lib    -Copt-level=3 file.rs`
### Error output

```
error[E0507]: cannot move out of `x` which is behind a mutable reference
  --> a.rs:10:18
   |
10 |     needs_fn_mut(async || {
   |                  ^^^^^^^^ `x` is moved here
11 |         x.hello();
   |         -
   |         |
   |         variable moved due to use in coroutine
   |         move occurs because `x` has type `Ty`, which does not implement the `Copy` trait
   |
note: if `Ty` implemented `Clone`, you could clone the value
  --> a.rs:15:1
   |
11 |         x.hello();
   |         - you could clone this value
...
15 | struct Ty;
   | ^^^^^^^^^ consider implementing `Clone` for this type
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at compiler/rustc_mir_transform/src/shim.rs:1082:13:
assertion `left matches right` failed: field should be captured by immutable ref if we have an `Fn` instance
  left: Ty
 right: ty::Ref(_, _, hir::Mutability::Not)
stack backtrace:
   0:     0x7d1d541c0fdd - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h8ceea53997a5b9b5
   1:     0x7d1d54a05297 - core::fmt::write::h05be396c0a785199
   2:     0x7d1d55994951 - std::io::Write::write_fmt::h0faf3273f924176d
   3:     0x7d1d541c36bb - std::panicking::default_hook::{{closure}}::h5ff21810d79cb0e4
   4:     0x7d1d541c332e - std::panicking::default_hook::h537184642066c310
   5:     0x7d1d53359419 - std[ddbe28eb9a61e65d]::panicking::update_hook::<alloc[a7f470ce4bd697bf]::boxed::Box<rustc_driver_impl[637e7f0798538e00]::install_ice_hook::{closure#0}>>::{closure#0}
   6:     0x7d1d541c3fd7 - std::panicking::rust_panic_with_hook::hc4a14d811dd29982
   7:     0x7d1d541c3c97 - std::panicking::begin_panic_handler::{{closure}}::h7df3c2502d8a1754
   8:     0x7d1d541c1499 - std::sys::backtrace::__rust_end_short_backtrace::h8861b37587c51d36
   9:     0x7d1d541c3964 - rust_begin_unwind
  10:     0x7d1d51094943 - core::panicking::panic_fmt::he0bc112117bdf86d
  11:     0x7d1d52d5cbef - core::panicking::assert_failed_inner::h5a404377cbdb79cc
  12:     0x7d1d53a6168b - core[b1645707a2a7518c]::panicking::assert_matches_failed::<&rustc_type_ir[66f9952f9eb239cc]::ty_kind::TyKind<rustc_middle[19486b3a4165f2e4]::ty::context::TyCtxt>>
  13:     0x7d1d55348106 - rustc_mir_transform[2c3e1838a62bac28]::shim::make_shim
  14:     0x7d1d5534512f - rustc_query_impl[9179f30695a188d8]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9179f30695a188d8]::query_impl::mir_shims::dynamic_query::{closure#2}::{closure#0}, rustc_middle[19486b3a4165f2e4]::query::erase::Erased<[u8; 8usize]>>
  15:     0x7d1d553450e3 - <rustc_query_impl[9179f30695a188d8]::query_impl::mir_shims::dynamic_query::{closure#2} as core[b1645707a2a7518c]::ops::function::FnOnce<(rustc_middle[19486b3a4165f2e4]::ty::context::TyCtxt, rustc_middle[19486b3a4165f2e4]::ty::instance::InstanceKind)>>::call_once
  16:     0x7d1d553447b5 - rustc_query_system[e02210f0baf32b51]::query::plumbing::try_execute_query::<rustc_query_impl[9179f30695a188d8]::DynamicConfig<rustc_query_system[e02210f0baf32b51]::query::caches::DefaultCache<rustc_middle[19486b3a4165f2e4]::ty::instance::InstanceKind, rustc_middle[19486b3a4165f2e4]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[9179f30695a188d8]::plumbing::QueryCtxt, false>
  17:     0x7d1d553444ff - rustc_query_impl[9179f30695a188d8]::query_impl::mir_shims::get_query_non_incr::__rust_end_short_backtrace
  18:     0x7d1d51556a9b - <rustc_middle[19486b3a4165f2e4]::ty::context::TyCtxt>::instance_mir
  19:     0x7d1d5555a65c - rustc_mir_transform[2c3e1838a62bac28]::inline::cycle::mir_inliner_callees
  20:     0x7d1d5555a528 - rustc_query_impl[9179f30695a188d8]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9179f30695a188d8]::query_impl::mir_inliner_callees::dynamic_query::{closure#2}::{closure#0}, rustc_middle[19486b3a4165f2e4]::query::erase::Erased<[u8; 16usize]>>
  21:     0x7d1d5555a4e7 - <rustc_query_impl[9179f30695a188d8]::query_impl::mir_inliner_callees::dynamic_query::{closure#2} as core[b1645707a2a7518c]::ops::function::FnOnce<(rustc_middle[19486b3a4165f2e4]::ty::context::TyCtxt, rustc_middle[19486b3a4165f2e4]::ty::instance::InstanceKind)>>::call_once
  22:     0x7d1d551993b3 - rustc_query_system[e02210f0baf32b51]::query::plumbing::try_execute_query::<rustc_query_impl[9179f30695a188d8]::DynamicConfig<rustc_query_system[e02210f0baf32b51]::query::caches::DefaultCache<rustc_middle[19486b3a4165f2e4]::ty::instance::InstanceKind, rustc_middle[19486b3a4165f2e4]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[9179f30695a188d8]::plumbing::QueryCtxt, false>
  23:     0x7d1d5519910d - rustc_query_impl[9179f30695a188d8]::query_impl::mir_inliner_callees::get_query_non_incr::__rust_end_short_backtrace
  24:     0x7d1d5596fe23 - rustc_mir_transform[2c3e1838a62bac28]::inline::cycle::mir_callgraph_reachable::process
  25:     0x7d1d5596f130 - rustc_mir_transform[2c3e1838a62bac28]::inline::cycle::mir_callgraph_reachable::process
  26:     0x7d1d5596e214 - rustc_mir_transform[2c3e1838a62bac28]::inline::cycle::mir_callgraph_reachable
  27:     0x7d1d5596e0a9 - rustc_query_impl[9179f30695a188d8]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9179f30695a188d8]::query_impl::mir_callgraph_reachable::dynamic_query::{closure#2}::{closure#0}, rustc_middle[19486b3a4165f2e4]::query::erase::Erased<[u8; 1usize]>>
  28:     0x7d1d5596e06b - <rustc_query_impl[9179f30695a188d8]::query_impl::mir_callgraph_reachable::dynamic_query::{closure#2} as core[b1645707a2a7518c]::ops::function::FnOnce<(rustc_middle[19486b3a4165f2e4]::ty::context::TyCtxt, (rustc_middle[19486b3a4165f2e4]::ty::instance::Instance, rustc_span[9cfca6d3892329e0]::def_id::LocalDefId))>>::call_once
  29:     0x7d1d55197fa5 - rustc_query_system[e02210f0baf32b51]::query::plumbing::try_execute_query::<rustc_query_impl[9179f30695a188d8]::DynamicConfig<rustc_query_system[e02210f0baf32b51]::query::caches::DefaultCache<(rustc_middle[19486b3a4165f2e4]::ty::instance::Instance, rustc_span[9cfca6d3892329e0]::def_id::LocalDefId), rustc_middle[19486b3a4165f2e4]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[9179f30695a188d8]::plumbing::QueryCtxt, false>
  30:     0x7d1d55197cf4 - rustc_query_impl[9179f30695a188d8]::query_impl::mir_callgraph_reachable::get_query_non_incr::__rust_end_short_backtrace
  31:     0x7d1d55365c45 - <rustc_mir_transform[2c3e1838a62bac28]::inline::Inliner>::try_inlining
  32:     0x7d1d553634e0 - <rustc_mir_transform[2c3e1838a62bac28]::inline::Inliner>::process_blocks
  33:     0x7d1d553629a6 - <rustc_mir_transform[2c3e1838a62bac28]::inline::Inline as rustc_middle[19486b3a4165f2e4]::mir::MirPass>::run_pass
  34:     0x7d1d54a02151 - rustc_mir_transform[2c3e1838a62bac28]::pass_manager::run_passes_inner
  35:     0x7d1d5555b537 - rustc_mir_transform[2c3e1838a62bac28]::optimized_mir
  36:     0x7d1d5554875d - rustc_query_impl[9179f30695a188d8]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9179f30695a188d8]::query_impl::optimized_mir::dynamic_query::{closure#2}::{closure#0}, rustc_middle[19486b3a4165f2e4]::query::erase::Erased<[u8; 8usize]>>
  37:     0x7d1d54a263e7 - rustc_query_system[e02210f0baf32b51]::query::plumbing::try_execute_query::<rustc_query_impl[9179f30695a188d8]::DynamicConfig<rustc_query_system[e02210f0baf32b51]::query::caches::DefIdCache<rustc_middle[19486b3a4165f2e4]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[9179f30695a188d8]::plumbing::QueryCtxt, false>
  38:     0x7d1d54a2599f - rustc_query_impl[9179f30695a188d8]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  39:     0x7d1d51556c44 - <rustc_middle[19486b3a4165f2e4]::ty::context::TyCtxt>::instance_mir
  40:     0x7d1d551a30c6 - rustc_interface[b9699dd6fb07a8c]::passes::run_required_analyses
  41:     0x7d1d5558671e - rustc_interface[b9699dd6fb07a8c]::passes::analysis
  42:     0x7d1d555866f1 - rustc_query_impl[9179f30695a188d8]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[9179f30695a188d8]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[19486b3a4165f2e4]::query::erase::Erased<[u8; 1usize]>>
  43:     0x7d1d5594fe2e - rustc_query_system[e02210f0baf32b51]::query::plumbing::try_execute_query::<rustc_query_impl[9179f30695a188d8]::DynamicConfig<rustc_query_system[e02210f0baf32b51]::query::caches::SingleCache<rustc_middle[19486b3a4165f2e4]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[9179f30695a188d8]::plumbing::QueryCtxt, false>
  44:     0x7d1d5594fb8f - rustc_query_impl[9179f30695a188d8]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  45:     0x7d1d557d8a6a - rustc_interface[b9699dd6fb07a8c]::interface::run_compiler::<core[b1645707a2a7518c]::result::Result<(), rustc_span[9cfca6d3892329e0]::ErrorGuaranteed>, rustc_driver_impl[637e7f0798538e00]::run_compiler::{closure#0}>::{closure#1}
  46:     0x7d1d557c0c90 - std[ddbe28eb9a61e65d]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[b9699dd6fb07a8c]::util::run_in_thread_with_globals<rustc_interface[b9699dd6fb07a8c]::util::run_in_thread_pool_with_globals<rustc_interface[b9699dd6fb07a8c]::interface::run_compiler<core[b1645707a2a7518c]::result::Result<(), rustc_span[9cfca6d3892329e0]::ErrorGuaranteed>, rustc_driver_impl[637e7f0798538e00]::run_compiler::{closure#0}>::{closure#1}, core[b1645707a2a7518c]::result::Result<(), rustc_span[9cfca6d3892329e0]::ErrorGuaranteed>>::{closure#0}, core[b1645707a2a7518c]::result::Result<(), rustc_span[9cfca6d3892329e0]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[b1645707a2a7518c]::result::Result<(), rustc_span[9cfca6d3892329e0]::ErrorGuaranteed>>
  47:     0x7d1d557c12fa - <<std[ddbe28eb9a61e65d]::thread::Builder>::spawn_unchecked_<rustc_interface[b9699dd6fb07a8c]::util::run_in_thread_with_globals<rustc_interface[b9699dd6fb07a8c]::util::run_in_thread_pool_with_globals<rustc_interface[b9699dd6fb07a8c]::interface::run_compiler<core[b1645707a2a7518c]::result::Result<(), rustc_span[9cfca6d3892329e0]::ErrorGuaranteed>, rustc_driver_impl[637e7f0798538e00]::run_compiler::{closure#0}>::{closure#1}, core[b1645707a2a7518c]::result::Result<(), rustc_span[9cfca6d3892329e0]::ErrorGuaranteed>>::{closure#0}, core[b1645707a2a7518c]::result::Result<(), rustc_span[9cfca6d3892329e0]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[b1645707a2a7518c]::result::Result<(), rustc_span[9cfca6d3892329e0]::ErrorGuaranteed>>::{closure#1} as core[b1645707a2a7518c]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  48:     0x7d1d557c166b - std::sys::pal::unix::thread::Thread::new::thread_start::hea7a4f91b7cfbbd5
  49:     0x7d1d56eee39d - <unknown>
  50:     0x7d1d56f7349c - <unknown>
  51:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/tmp/im/rustc-ice-2024-08-19T05_31_11-2769501.txt` to your bug report

note: compiler flags: -Z validate-mir --crate-type lib -C opt-level=3

query stack during panic:
#0 [mir_shims] generating MIR shim for `hello::{closure#0}`
#1 [mir_inliner_callees] computing all local function calls in `hello::{closure#0}`
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0507`.
```

</p>
</details>

--- a/compiler/rustc_mir_transform/src/shim.rs
+++ b/compiler/rustc_mir_transform/src/shim.rs
@@ -1070,19 +1070,26 @@ fn build_construct_coroutine_by_move_shim<'tcx>(
     let locals = local_decls_for_sig(&sig, span);
 
     let mut fields = vec![];
+
+    // Move all of the closure args.
     for idx in 1..sig.inputs().len() {
         fields.push(Operand::Move(Local::from_usize(idx + 1).into()));
     }
+
     for (idx, ty) in args.as_coroutine_closure().upvar_tys().iter().enumerate() {
         if receiver_by_ref {
             // The only situation where it's possible is when we capture immuatable references,
             // since those don't need to be reborrowed with the closure's env lifetime. Since
             // references are always `Copy`, just emit a copy.
-            assert_matches!(
-                ty.kind(),
-                ty::Ref(_, _, hir::Mutability::Not),
-                "field should be captured by immutable ref if we have an `Fn` instance"
-            );
+            if !matches!(ty.kind(), ty::Ref(_, _, hir::Mutability::Not)) {
+                // This copy is only sound if it's a `&T`. This may be
+                // reachable e.g. when eagerly computing the `Fn` instance
+                // of an async closure that doesn't borrowck.
+                tcx.dcx().delayed_bug(format!(
+                    "field should be captured by immutable ref if we have \
+                    an `Fn` instance, but it was: {ty}"
+                ));
+            }
             fields.push(Operand::Copy(tcx.mk_place_field(
                 self_local,
                 FieldIdx::from_usize(idx),

--- a/tests/ui/async-await/async-closures/closure-shim-borrowck-error.rs
+++ b/tests/ui/async-await/async-closures/closure-shim-borrowck-error.rs
@@ -1,4 +1,3 @@
-//@ known-bug: rust-lang/rust#129262
 //@ compile-flags: -Zvalidate-mir --edition=2018 --crate-type=lib -Copt-level=3
 
 #![feature(async_closure)]
@@ -11,6 +10,7 @@ fn needs_fn_mut<T>(mut x: impl FnMut() -> T) {
 
 fn hello(x: Ty) {
     needs_fn_mut(async || {
+        //~^ ERROR cannot move out of `x`
         x.hello();
     });
 }

--- a/tests/ui/async-await/async-closures/closure-shim-borrowck-error.stderr
+++ b/tests/ui/async-await/async-closures/closure-shim-borrowck-error.stderr
@@ -1,0 +1,24 @@
+error[E0507]: cannot move out of `x` which is behind a mutable reference
+  --> $DIR/closure-shim-borrowck-error.rs:12:18
+   |
+LL |     needs_fn_mut(async || {
+   |                  ^^^^^^^^ `x` is moved here
+LL |
+LL |         x.hello();
+   |         -
+   |         |
+   |         variable moved due to use in coroutine
+   |         move occurs because `x` has type `Ty`, which does not implement the `Copy` trait
+   |
+note: if `Ty` implemented `Clone`, you could clone the value
+  --> $DIR/closure-shim-borrowck-error.rs:18:1
+   |
+LL |         x.hello();
+   |         - you could clone this value
+...
+LL | struct Ty;
+   | ^^^^^^^^^ consider implementing `Clone` for this type
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0507`.

